ibis/expr/operations.py

```diff
@@ -349,6 +349,17 @@ def output_type(self):
             return rules.shape_like(arg, 'double')
 
 
+class Clip(ValueOp):
+    input_type = [number(name='lower', allow_boolean=False, optional=True),
+                  number(name='upper', allow_boolean=False, optional=True)]
+
+    def output_type(self):
```

Comment:
The output type here should be the same as the input type. You can use `rules.type_of_arg(0)` to get the output type of the first input argument. You should also be able to just assign that directly to `output_type` like this:

```python
class FooNode(...):
    output_type = rules.type_of_arg(0)
```